### ISSUE TITLE
Reducer bucketing based on autograd profile 

### ISSUE THREAD

**Issue by pietern** [pytorch/pytorch#22340]
## 🚀 Feature

How gradient tensors are batched together in bucket prior to reducing them across processes is done completely statically today. A user specifies a lower bound on the bucket size and parameters are bucketed accordingly. We can use profile data of the backwards pass to find a bucket assignment that better aligns with the point in time gradients are ready.

## Motivation

In some vision models the last layers have many weights (big fully connected layers) and the first layers have only a few weights (small convolutional layers) The bulk of gradients are computed early during the backwards pass. Conversely the gradient tensors for the first layers are computed last and are relatively small. If the final bucket to reduce is 25MB and we wait for the final 1KB worth of gradients before we start, we may achieve better end to end latency if we instead start reduction sooner, without the final 1KB, and run a separate reduction for the final piece of data. We have hooks in `c10d::Reducer`, as well as the autograd profiler, both of which can be used to generate timing data for the gradient of every parameter. Depending on this timing and the size of the gradients, we can create a bucketing assignment that is a better match for the execution order of autograd and doesn't have a big "latency bubble" after autograd has finished. 

## Pitch

* Evaluate applicability of the autograd profiler here. E.g. can we get a mapping of the indices in `model.parameters()` to some relative timestamp their gradient was computed? If not, how easy would it be to add it? (If the autograd profiler cannot be used we can look into adding timing to `c10d::Reducer` directly.)
* Generate a new bucketing assignment based on relative time per parameter as well as the size of the corresponding gradient.
* Distribute the newly computed bucketing order to every participating process. This is done by nvidia/apex as well (cc @mcarilli) where the parameter order is distributed to all processes. If there is mismatch between the order that the model layers/parameters are defined and the order they are used, any approach that closely resembles the autograd order will already be an improved over the naive strategy of simply using definition order.  

## Additional context

Stretch goal: generate a number of bucketing assignments and then empirically figure out which one is best for the set of machines the model is running on.
